mldsa/reduce.h

```diff
@@ -41,9 +52,40 @@ __contract__(
 );
 
 #define caddq MLD_NAMESPACE(caddq)
-int32_t caddq(int32_t a);
+/*************************************************
+ * Name:        caddq
+ *
+ * Description: Add MLDSA_Q if input coefficient is negative.
+ *
+ * Arguments:   - int32_t: finite field element a
+ *
+ * Returns r.
+ **************************************************/
+int32_t caddq(int32_t a)
+__contract__(
+  requires(a > -MLDSA_Q)
+  requires(a < MLDSA_Q)
+  ensures(return_value >= 0)
+  ensures(return_value < MLDSA_Q)
+  ensures(return_value == (a >= 0) ? a : (a + MLDSA_Q))
+);
 
 #define freeze MLD_NAMESPACE(freeze)
-int32_t freeze(int32_t a);
+/*************************************************
+ * Name:        freeze
+ *
+ * Description: For finite field element a, compute standard
+ *              representative r = a mod^+ MLDSA_Q.
+ *
+ * Arguments:   - int32_t: finite field element a
+ *
+ * Returns r.
+ **************************************************/
+int32_t freeze(int32_t a)
+__contract__(
+  requires(a <= REDUCE_DOMAIN_MAX)
+  ensures(return_value >= 0)
+  ensures(return_value < MLDSA_Q)
```

Comment:
Can we add a cassert at the end of the function body that the return value is actually congruent to a mod q?

Reply:
Possibly, but I can't work out how to express that using C's % operator...
We'd also need a stronger ensures contract on the spec of reduce32()

Reply:
`cassert((t - a) % MLDSA_Q == 0);`?
that works for `reduce32`

Reply:
ahhh but of course you need to add that to the contracts of `caddq` and `reduce32`

Reply:
Are we really going to need a strong postcondition on freeze()? Not sure...

Reply:
I can't find a single call to freeze() anywhere.... perhaps we should just delete it?!?

Reply:
Very good point. Please delete it.